### PR TITLE
First bits of wgpu-remote

### PR DESCRIPTION
**Pull request by kvark**

### CODE REVIEW

File: wgpu-remote/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    devices: wgn::IdentityManager,
+}
+
+pub struct Client {
+    channel: IpcSender<GlobalMessage>,
+    identity: Mutex<IdentityHub>,
+}
+
+impl Client {
+    pub fn new(channel: IpcSender<GlobalMessage>) -> Self {
```
Comment (grovesNL) [inline wgpu-remote/src/lib.rs]: We'll need to expose this in a C extern too, right?

File: wgpu-bindings/wgpu-remote.h [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+typedef struct WGPUClient WGPUClient;
+
+WGPUDeviceId wgpu_adapter_create_device(const WGPUClient *client,
```
[the remark this answers is not in the document]
Comment (kvark) [reply wgpu-bindings/wgpu-remote.h]: Probably. We also don't expect the same application to link to both remote and native.

File: wgpu-native/src/hub.rs [hunk 3]
```diff
@@ -7,56 +7,89 @@ use crate::{
...
-//TODO: use Vec instead of HashMap
-//TODO: track epochs of indices
 
-pub type Id = u32;
+pub(crate) type Index = u32;
+pub(crate) type Epoch = u32;
+#[derive(Clone, Copy, Debug, Hash, PartialEq)]
+pub struct Id(Index, Epoch);
+
+pub trait NewId {
```
[the remark this answers is not in the document]
Comment (kvark) [reply wgpu-native/src/hub.rs]: because I wanted to keep `Index` and `Epoch` fields private.

File: wgpu-remote/src/lib.rs (hunk 1 above)
Comment (kvark) [reply wgpu-remote/src/lib.rs]: sure, at some point in the near future :)

**Comment by kvark** [comment]
@grovesNL 
> Could you also explain more about how epochs will be used with resource tracking, especially how they relate to IDs here?

Sure! Basically speaking, in most places an ID is a pair of index and epoch. Whenever we need to consider this API to be a part of the state, we check the epoch with one in the state. Never does the code require to keep track of the same ID under multiple epochs. Therefore, if there is a `Map<Id, Something>`, it's always an improvement to move the epoch from the key side to the value side, making it `Map<Index, (Something, Epoch)>`. This allows us to quickly catch any epoch mismatches, and also reduces the key size, which makes hashing a little bit faster.

File: wgpu-bindings/wgpu-remote.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (grovesNL) [reply wgpu-bindings/wgpu-remote.h]: I'm interested in this mostly to reduce confusion while reviewing or searching code :)

**Comment by kvark** [comment]
Rebased now
bors r=grovesNL

**Pull request merged**